disnake/enums.py

```diff
@@ -72,326 +85,1370 @@
+
+        cls = super().__new__(metacls, name, bases, ns)
+
+        for name_, value_ in namespace.member_map.items():
+            member = cls.__new__(cls, value_)  # type: ignore
+            member._name_ = name_
+            member._value_ = value_
+            name_map[name_] = member
+            value_map.setdefault(value_, member)  # Prioritize first defined in case of alias
+            setattr(cls, name_, member)
+
+        return cls
+
+    @classmethod
+    def __prepare__(
+        metacls, name: str, bases: Tuple[Type[Any], ...] = (), /, **kwds: Any
+    ) -> Union[Dict[str, Any], _EnumDict]:
+        # with this we get to ensure the new class' namespace is an _EnumDict
+
+        if not EnumMeta.__is_enum_instantiated:
+            return _EnumDict(object)
+
+        try:
+            base, _ = bases  # 'loss' of functionality: only (type, Enum) enums are allowed
+        except ValueError:
+            raise TypeError("Expected exactly two base classes for an enum") from None
+
+        if isinstance(base, EnumMeta):
+            raise TypeError("An Enum's first base class must be its values' type")
+
+        return _EnumDict(base)
+
+    def __repr__(cls) -> str:
         return f"<enum {cls.__name__}>"
 
+    def __call__(cls, value: Any) -> Any:
+        try:
+            return cls._value2member_map_[value]
+        except KeyError:
+            raise ValueError(f"{value} is not a valid {cls.__name__}") from None
+
+    def __getitem__(cls, name: str) -> Any:
+        return cls._name_map_[name]
+
+    def __contains__(cls, value: Any) -> bool:
+        return value in cls._value2member_map_
+
+    def __iter__(cls) -> Iterator[Any]:
+        yield from cls._name_map_.values()
+
     @property
-    def __members__(cls):
-        return types.MappingProxyType(cls._enum_member_map_)
+    def __members__(cls) -> Mapping[str, Enum]:
+        return types.MappingProxyType(cls._name_map_)
 
-    def __call__(cls, value):
-        try:
-            return cls._enum_value_map_[value]
-        except (KeyError, TypeError):
-            raise ValueError(f"{value!r} is not a valid {cls.__name__}")
+    @property
+    def _member_names_(cls) -> Sequence[str]:
+        # I *think* this should be fine as a property?
+        # Hardly ever gets used so I don't really see the value in pre-computing it like
+        # vanilla Enums do. I decided to save memory but we can always just revert this.
+        return tuple(cls._name_map_)
+
+
+if TYPE_CHECKING:
+    from enum import Enum
+else:
+
+    class Enum(metaclass=EnumMeta):
+        _name_map_: ClassVar[Mapping[str, Enum]]
+        _value2member_map_: ClassVar[Mapping[Any, Enum]]
+        _name_: str
+        _value_: Any
+
+        @property
+        def name(self) -> str:
+            """The name of the enum member."""
+            return self._name_
+
+        @property
+        def value(self):
+            """The value of the enum member."""
+            return self._value_
```

Comment:
Pretty much a shot in the dark, as I'm guessing some thought has already gone into this, but - is there a way to rewrite the implementation to avoid properties here?
Since `.value` is still being used everywhere for now, this currently comes with a 200% performance penalty compared to the previous namedtuple access (101ns vs 31.8ns). Certainly not critical at this scale, but still something I noticed while reading through this.

Reply:
Yeah, avoiding properties is as easy as removing the property and changing `_value_` to `value` both here and in `EnumMeta.__new__`. I chose to use properties to make it harder for a user to accidentally overwrite a disnake enum value and have their bot stop functioning properly, but it's slightly far-fetched. I'm personally quite torn on this, which would you prefer?

Reply:
Yeah, thought about just renaming them as well, but keeping them read-only is more important than avoiding properties here.
I wonder if `operator.attrgetter` would be faster, as it's implemented natively... i'm guessing it won't do much (if anything), so keeping it like this would be fine too.

Reply:
Did some more testing, looks like `attrgetter` may still come in handy: there doesn't seem to be a way to avoid the call overhead with builtin tools, but native calls are still a bit faster than pure python calls.
Some benchmarks (this time properly, with pyperf and more inner loops lol):
```rb
# namedtuple attribute access (old impl)
$ pyperf timeit -q --rigorous --duplicate 512 -s 'import disnake; c = disnake.ChannelType.text' 'c.name'
Mean +- std dev: 15.9 ns +- 1.3 ns

# property (current impl)
$ pyperf timeit -q --rigorous --duplicate 512 -s 'import disnake; c = disnake.ChannelType.text' 'c.name'
Mean +- std dev: 73.3 ns +- 7.6 ns

# property + attrgetter
$ pyperf timeit -q --rigorous --duplicate 512 -s 'import disnake; c = disnake.ChannelType.text' 'c.name'
Mean +- std dev: 48.9 ns +- 4.0 ns
```


```diff
diff --git a/disnake/enums.py b/disnake/enums.py
index 14036ae3..cef23c41 100644
--- a/disnake/enums.py
+++ b/disnake/enums.py
@@ -23,6 +23,7 @@ FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 DEALINGS IN THE SOFTWARE.
 """
 from __future__ import annotations
+from operator import attrgetter
 
 import types
 from typing import (
@@ -232,10 +233,7 @@ else:
         _name_: str
         _value_: Any
 
-        @property
-        def name(self) -> str:
-            """The name of the enum member."""
-            return self._name_
+        name = property(attrgetter("_name_"), doc="The name of the enum member.")
 
         @property
         def value(self):

```

Reply:
Looks rather hacky but since it retains functionality and still offers a speedup, I think that may just be worth it. I'll keep this in mind ^^

Reply:
Implemented as part of 662e0bdf1f0c47a3142d4214b1cbda16fb6e5b17.

Reply:
This was changed to direct access with a special `__setattr__` to protect `.name` and `.value` as part of 057f09b51e3ba2bc2fff37065f3072c0c940e378. The extra check can also be reverted again, which was done in 31757782843f0e01c9f3598b223faa8274f73264.

This is considerably faster than even the `property(attrgetter(...))` hack, and still retains the security. Thanks to EQUENOS for the suggestion ^^